### PR TITLE
Fix e2e tests for providers without SSH to the nodes

### PR DESCRIPTION
**Pull request by sttts**
A number of e2e tests which use SSH are skipped for providers other than gce, gke and aws. This patch does the same for 
- the kubeproxy tests (only for the probes which actually use SSH, the others are kept running)
- "should release NodePorts on delete" in service.go
- "should test privileged pod" in privileged.go.

Moreover, the provider list of those which support SSH is factored out into `providersWithSSH`.

Fixes https://github.com/mesosphere/kubernetes-mesos/issues/500

### CODE REVIEW

File: test/e2e/kubeproxy.go [hunk 1]
```diff
@@ -417,13 +428,32 @@ func (config *KubeProxyTestConfig) setup() {
...
+			for _, a := range nodes.Items[i].Status.Addresses {
+				if a.Type == api.NodeLegacyHostIP {
+					config.nodes = append(config.nodes, a.Address)
+					continue NodeLoop
+				}
+			}
+		}
+	}
+	if len(config.nodes) == 0 {
+		Failf("No nodes with externally reachable address found")
```
Comment (s-urbaniak) [inline test/e2e/kubeproxy.go]: `Failf` implies the whole test must fail, is this true?

File: test/e2e/service.go [hunk 2]
```diff
@@ -532,14 +532,17 @@ var _ = Describe("Services", func() {
...
-			Failf("expected node port (%d) to be in use", nodePort)
+		// this test uses NodeSSHHosts that does not work if a Node only reports LegacyHostIP
+		if providerIs("gce", "gke", "aws") {
+			hosts, err := NodeSSHHosts(c)
+			if err != nil {
+				Expect(err).NotTo(HaveOccurred())
+			}
+			cmd := fmt.Sprintf(`test -n "$(ss -ant46 'sport = :%d' | tail -n +2 | grep LISTEN)"`, nodePort)
+			_, _, code, err := SSH(cmd, hosts[0], testContext.Provider)
+			if code != 0 {
```
Comment (s-urbaniak) [inline test/e2e/service.go]: Can `code == 0 && err != nil` happen? if yes, we should check for `err != nil`?

File: test/e2e/kubeproxy.go (hunk 1 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: yes, fatal error.

File: test/e2e/service.go (hunk 2 above)
Comment (sttts) [inline test/e2e/service.go]: probably. But I this is kind of out of scope of this PR. The code is the old code, just indented.

**Comment by s-urbaniak** [comment]
LGTM

**Comment by jdef** [comment]
rather than copy/pasting `providerIs("gce", "gke", "aws")` in several places, is it worth adding a top-level func `ProviderSupportsSSH() bool` that contains this logic?

**Comment by sttts** [comment]
@jdef I like that idea.

**Comment by sttts** [comment]
Rebased and squashed.

**Comment by sttts** [comment]
@s-urbaniak PTAL, rebased and squashed.

**Comment by s-urbaniak** [comment]
LGTM

**Comment by sttts** [comment]
@ArtfulCoder I have assigned this to you for a review as you wrote most of the kubeproxy e2e tests.

File: test/e2e/kubeproxy.go [hunk 3]
```diff
@@ -60,6 +61,7 @@ type KubeProxyTestConfig struct {
 	nodePortService     *api.Service
 	loadBalancerService *api.Service
 	nodes               []string
+	enableSSHTests      bool
```
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: where is this variable used ?

File: test/e2e/kubeproxy.go [hunk 4]
```diff
@@ -38,7 +40,6 @@ import (
 const (
 	endpointHttpPort        = 8080
 	endpointUdpPort         = 8081
-	endpointHostPort        = 8082
```
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: why was this removed ?

also, if you have good reason to remove it, we should do it in a separate PR..

File: test/e2e/kubeproxy.go [hunk 5]
```diff
@@ -225,13 +228,17 @@ func (config *KubeProxyTestConfig) dialFromContainer(protocol, containerIP, targ
...
 		tries)
 
-	By(fmt.Sprintf("Dialing from container. Running command:%s", cmd))
-	stdout := config.ssh(cmd)
-	var output map[string][]string
-	err := json.Unmarshal([]byte(stdout), &output)
-	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Could not unmarshal curl response: %s", stdout))
-	hostNamesMap := array2map(output["responses"])
-	Expect(len(hostNamesMap)).To(BeNumerically("==", expectedCount), fmt.Sprintf("Response was:%v", output))
+	if providerIs(providersWithSSH...) {
```
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: I think we should use kubectl exec into a net=host pod and run the shell commands to work around the no-ssh-available problem.
For purposes of this PR, I would just run the entire test  for providerWithSSH, otherwise skip the whole test.
We are skipping pretty much everything other than the LoadBalancer portion..

File: test/e2e/kubeproxy.go [hunk 6]
```diff
@@ -417,13 +423,32 @@ func (config *KubeProxyTestConfig) setup() {
...
+	config.nodes = []string{}
+	if providerIs(providersWithSSH...) {
+		By("Getting ssh-able hosts")
+		hosts, err := NodeSSHHosts(config.f.Client)
+		Expect(err).NotTo(HaveOccurred())
+		for _, h := range hosts {
+			config.nodes = append(config.nodes, strings.TrimSuffix(h, ":22"))
+		}
+	} else {
+		// fall back to legacy IP
```
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: can this logic be in the NodeSSHHosts function ?

File: test/e2e/kubeproxy.go (hunk 5 above)
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: slightly off-topic:
In general, we want to use this container to run shell commands on the node.
 https://github.com/kubernetes/kubernetes/tree/master/contrib/for-tests/netexec 
We can use kubectl to hit this container using kubectl exec..

File: test/e2e/kubeproxy.go (hunk 3 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: Left over from a refactoring. enableSSHTests must go.

File: test/e2e/kubeproxy.go (hunk 4 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: It had been in its own commit before squashing. Will extract it again.

About the why: it doesn't seem to be used.

File: test/e2e/kubeproxy.go (hunk 5 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: I like the idea of a net=host pod. net=host pods are disabled though AFAIK for every other pod config source than file. This can be overwritten by a kubelet flag. So you have to configure your cluster for net=host support.

Thoughts?

File: test/e2e/kubeproxy.go (hunk 6 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: It does not make sense in general I think, only for this test. But I removed this now, following your comment above to skip the whole test. I agree a net=host pod is the way to go.

**Comment by sttts** [comment]
@ArtfulCoder addressed all your comments.

What do you think about the net=host question, i.e. that it requires either a static pod or a different kubelet configuration?

File: test/e2e/kubeproxy.go (hunk 5 above)
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: yes, i remember now 
i had worked on that file-based net=host pod feature :)
I guess, there isnt an easy way to do that yet..

File: test/e2e/kubeproxy.go (hunk 4 above)
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: it is required to  ensure that the pods dont share the same host.

File: test/e2e/kubeproxy.go [hunk 7]
```diff
@@ -242,13 +248,17 @@ func (config *KubeProxyTestConfig) dialFromNode(protocol, targetIP string, targe
 		cmd = fmt.Sprintf("curl -s --connect-timeout 1 http://%s:%d/hostName", targetIP, targetPort)
 	}
 	forLoop := fmt.Sprintf("for i in $(seq 1 %d); do %s; echo; done | grep -v '^\\s*$' |sort | uniq -c | wc -l", tries, cmd)
-	By(fmt.Sprintf("Dialing from node. command:%s", forLoop))
-	stdout := config.ssh(forLoop)
-	Expect(strconv.Atoi(strings.TrimSpace(stdout))).To(BeNumerically("==", expectedCount))
+	if providerIs(providersWithSSH...) {
```
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: you wont need the if now..

File: test/e2e/kubeproxy.go [hunk 8]
```diff
@@ -284,11 +294,6 @@ func (config *KubeProxyTestConfig) createNetShellPodSpec(podName string) *api.Po
 							Name:          "udp",
 							ContainerPort: endpointUdpPort,
 						},
-						{
-							Name:          "host",
```
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: need this back in

File: test/e2e/kubeproxy.go [hunk 9]
```diff
@@ -420,10 +425,10 @@ func (config *KubeProxyTestConfig) setup() {
 	By("Getting ssh-able hosts")
 	hosts, err := NodeSSHHosts(config.f.Client)
 	Expect(err).NotTo(HaveOccurred())
-	if len(hosts) == 0 {
```
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: can we add this check back

File: test/e2e/kubeproxy.go [hunk 10]
```diff
@@ -420,10 +425,10 @@ func (config *KubeProxyTestConfig) setup() {
...
-		Failf("No ssh-able nodes")
+	config.nodes = make([]string, len(hosts))
+	for _, h := range hosts {
+		config.nodes = append(config.nodes, strings.TrimSuffix(h, ":22"))
```
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: can we append :22 when we actually make the ssh call

File: test/e2e/kubeproxy.go (hunk 5 above)
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: you wont need the if now

File: test/e2e/kubeproxy.go (hunk 10 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: we do, see config.ssh

File: test/e2e/kubeproxy.go (hunk 4 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: Oh, then this deserves a comment ;-)

Comment (sttts) [inline test/e2e/kubeproxy.go]: Would you mind to set the NodeName correctly for each pod instead of this "port hack". Constant host ports are not nice.

File: test/e2e/kubeproxy.go (hunk 10 above)
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: yes, thanks

File: test/e2e/kubeproxy.go (hunk 4 above)
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: its a fair suggestion. I would keep the hostPort and do the NodeName in a separate PR.
Lets not combine things in the same PR.

File: test/e2e/kubeproxy.go (hunk 9 above)
Comment (sttts) [inline test/e2e/kubeproxy.go]: NodeSSHHosts checks this:
```
if len(hosts) != len(nodelist.Items) {
```

We won't run this test on an empty cluster I guess.

**Comment by sttts** [comment]
@ArtfulCoder ptal. I have some more patches in the pipe, e.g. NodeName based scheduling, parallel startup of the pod and a prototype of kubectl exec instead of ssh. Will create new PRs later for that.

**Comment by sttts** [comment]
Rebased and squashed.

File: test/e2e/kubeproxy.go (hunk 9 above)
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: We won't run the test and the test will appear to be green when it was in-fact not run.
So, this check basically says that: something is wrong.

**Comment by ArtfulCoder** [comment]
for the hostNetwork=true question, I have a PR in flight to allow that for non-file-based pods as well.
#15021

**Comment by ArtfulCoder** [comment]
@sttts 
Basically, even if we cannot SSH on to nodes, we can use a pod with hostNetwork = true (post- #15021) and run shell commands from the pod.we can use privileged=true if required.(#13930)

File: test/e2e/kubeproxy.go (hunk 9 above)
Comment (ArtfulCoder) [inline test/e2e/kubeproxy.go]: sounds good.

**Comment by sttts** [comment]
@ArtfulCoder very cool about allowing net=host by default. Let me push my port to that next week. Will also take a look at the other ssh using e2e tests. Getting rid of ssh would be great.

**Comment by sttts** [comment]
Shippable 1.4 times out, 1.3 is green. Rebased. Let's see.
